### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_df = pd.read_csv(file_to_load)
purchase_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Display the total number of players


In [2]:
#Total number of players
#Calculate the total number of players by using SN as it is the unique value
player=len(purchase_df["SN"].value_counts())

#Create summary data frame
player_total=pd.DataFrame([player], columns = ["Total Players"])

player_total

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [17]:
#Calculate the reqs
#No. of Unique Items
unique_total=len(purchase_df["Item ID"].unique())

#Average Price
average=round(purchase_df["Price"].mean(),2)

#No. of Purchases
purchases=len(purchase_df["Purchase ID"].unique())

#Total Revenue
revenue=purchase_df["Price"].sum()

#Create summary data frame
#Use the calculated data and create the columns with labels
analysis=pd.DataFrame({"Number of Unique Items":[unique_total], "Average Price":[average], "Number of Purchases":[purchases], "Total Revenue":[revenue]})
prices=['Average Price', 'Total Revenue']

#Format the columns with prices using $X.XX format 
analysis[prices]=analysis[prices].applymap(lambda x:'${:.2f}'.format(x))

analysis

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     183         $3.05                  780      $2379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [4]:
#Create a new df to remove duplicated SN
gender=purchase_df[["SN","Gender"]]

#Remove the duplicates in the list
gender=gender.drop_duplicates()

#Count the players per gender
gender_count=gender["Gender"].value_counts()

#Getting the data for the demographics table
values=[gender_count[0], gender_count[1], gender_count[2]]
percentages=[(gender_count[0]/player)*100,(gender_count[1]/player)*100,(gender_count[2]/player)*100]

#Creating the DF
#Use the calculated data and create the columns and indeces with labels
demo_table = pd.DataFrame({"Total Count":values,"Percentage of Players":percentages})
demo_table.index=(["Male", "Female", "Other / Non-Disclosed"])

#Format the columns with prices using $X.XX format 
demo_table['Percentage of Players']=demo_table['Percentage of Players'].map('{:.2f}%'.format)

demo_table


Total Count Percentage of Players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [18]:
#Create the data
#Using groupby Gender and calculating the reqs
group=purchase_df.groupby(["Gender"])
pur_count=group["SN"].count()
pur_avg=group["Price"].mean()
pur_tot=group["Price"].sum()
pur_person=pur_tot/pur_count

#Creating the DF
#Use the calculated data and create the columns with labels
pur_gender = pd.DataFrame({"Purchase Count": pur_count, "Average Purchase Price": pur_avg,"Total Purchase Value": pur_tot,
                              "Avg Total Purchase per Person": pur_person})
gender_prices=['Average Purchase Price', 'Total Purchase Value','Avg Total Purchase per Person']

#Format the columns with prices using $X.XX format 
pur_gender[gender_prices]=pur_gender[gender_prices].applymap(lambda x:'${:.2f}'.format(x))

pur_gender


Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $3.20  
Male                              $1967.64                         $3.02  
Other / Non-Disclosed               $50.19                         $3.35

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [19]:
#Create an new df to remove duplicated SN but using Age data
nondup_df=purchase_df[["SN","Age"]]

#Remove the duplicates
nondup_df=nondup_df.drop_duplicates()

#Create bins per age group
bins=[0,9,14,19,24,29,34,39,200]

#Create labels for the bins
grp_lbl=["<10","10-14","15-19","20-24","25-29","30-34","35-39","40+"]

#Slice the data and place them into each bin
pd.cut(nondup_df["Age"],bins,labels=grp_lbl).head()

#Place the data series into a new column inside of the DF
nondup_df["Age Ranges"]=pd.cut(nondup_df["Age"],bins,labels=grp_lbl)

#Create a Groupby object based on "Age Ranges"
age_range=nondup_df.groupby("Age Ranges")

#Find how many players in each range/bin and its percentages
tot_count=age_range["Age"].count()
percentage=tot_count/player*100

#Creating the DF
#Use the calculated data and create the columns with labels
agedemo_df=pd.DataFrame({"Total Count":tot_count, "Percentage of Players":percentage})

#Format the columns with prices using $X.XX format 
agedemo_df['Percentage of Players']=agedemo_df['Percentage of Players'].map('{:.2f}%'.format)

agedemo_df


Total Count Percentage of Players
Age Ranges                                   
<10                  17                 2.95%
10-14                22                 3.82%
15-19               107                18.58%
20-24               258                44.79%
25-29                77                13.37%
30-34                52                 9.03%
35-39                31                 5.38%
40+                  12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [7]:
purchase_df["Age Ranges"]=pd.cut(purchase_df["Age"],bins,labels=grp_lbl)

#Create a Groupby object based on "Age Ranges"
age_range=purchase_df.groupby("Age Ranges")

#Calculate the reqs
age_purct=age_range["Purchase ID"].count()
age_prcavg=age_range["Price"].mean()
age_purtot=age_range["Price"].sum()
age_purper=age_purtot/tot_count

#Create the DF
#Create new columns and use the calculated reqs data
age_analysis=pd.DataFrame({"Purchase Count":age_purct,"Average Purchase Price":age_prcavg,"Total Purchase Value":age_purtot,"Avg Total Purchase per Person":age_purper})

#Format the columns with prices using $X.XX format and label the columns
age_prices=['Average Purchase Price', 'Total Purchase Value','Avg Total Purchase per Person']
age_analysis[age_prices]=age_analysis[age_prices].applymap(lambda x:'${:.2f}'.format(x))

age_analysis


Purchase Count Average Purchase Price Total Purchase Value  \
Age Ranges                                                               
<10                     23                  $3.35               $77.13   
10-14                   28                  $2.96               $82.78   
15-19                  136                  $3.04              $412.89   
20-24                  365                  $3.05             $1114.06   
25-29                  101                  $2.90              $293.00   
30-34                   73                  $2.93              $214.00   
35-39                   41                  $3.60              $147.67   
40+                     13                  $2.94               $38.24   

           Avg Total Purchase per Person  
Age Ranges                                
<10                                $4.54  
10-14                              $3.76  
15-19                              $3.86  
20-24                              $4.32  
25-29                              $3.81  
30-34                              $4.12  
35-39                              $4.76  
40+                                $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [14]:
#Create a Groupby object based on SN
spender=purchase_df.groupby("SN")

#Calculate the reqs
sn_purct=spender["Purchase ID"].count()
sn_prcavg=spender["Price"].mean()
sn_purtot=spender["Price"].sum()

#Create the DF
#Create new columns and use the calculated reqs data
spender_df=pd.DataFrame({"Purchase Count":sn_purct,"Average Purchase Price":sn_prcavg,"Total Purchase Value":sn_purtot})

#Sort with from the highest Total Purchase Value
spender_df=spender_df.sort_values("Total Purchase Value", ascending=False)

#Format the columns with prices using $X.XX format and label the columns
spender_prices=['Average Purchase Price', 'Total Purchase Value']
spender_df[spender_prices]=spender_df[spender_prices].applymap(lambda x:'${:.2f}'.format(x))

spender_df.head()


Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [11]:
#Create a Groupby object
popular=purchase_df.groupby(['Item ID', 'Item Name'])['Price'].agg(['count','mean','sum'])
popular.columns=['Purchase Count','Average Purchase Price','Total Purchase Value']

#Sort according to descending Purchase Count
popular_df=popular.sort_values("Purchase Count", ascending=False)

#Format the columns with prices using $X.XX format and label the columns
pop_prices=['Average Purchase Price', 'Total Purchase Value']
popular_df[pop_prices]=popular_df[pop_prices].applymap(lambda x:'${:.2f}'.format(x))

popular_df.head()


Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
19      Pursuit, Cudgel of Necromancy                              8   

                                                     Average Purchase Price  \
Item ID Item Name                                                             
178     Oathbreaker, Last Hope of the Breaking Storm                  $4.23   
145     Fiery Glass Crusader                                          $4.58   
108     Extraction, Quickblade Of Trembling Hands                     $3.53   
82      Nirvana                                                       $4.90   
19      Pursuit, Cudgel of Necromancy                                 $1.02   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
19      Pursuit, Cudgel of Necromancy                               $8.16

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [15]:
#Use the same dataframe "popular"
profit_df=popular.sort_values('Total Purchase Value',ascending=False)

#Format the columns with prices using $X.XX format and label the columns
profit_df[pop_prices]=profit_df[pop_prices].applymap(lambda x:'${:.2f}'.format(x))

profit_df.head()

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
92      Final Critic                                               8   
103     Singed Scalpel                                             8   

                                                     Average Purchase Price  \
Item ID Item Name                                                             
178     Oathbreaker, Last Hope of the Breaking Storm                  $4.23   
82      Nirvana                                                       $4.90   
145     Fiery Glass Crusader                                          $4.58   
92      Final Critic                                                  $4.88   
103     Singed Scalpel                                                $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
92      Final Critic                                               $39.04  
103     Singed Scalpel                                             $34.80

## Observations

Males purchased six times more than females which is directly proportional to the number of males against females.  This yields to the most Total Purchased Value as well.  However, it is noted that each male purhcased cheaper as compared to each female.  Those players who did not declare their gender, purchased the most values per player/ person.

The most number of purchases came from the ages 20-24 which comprises of 45% of the total purchases.  This also yields the highest Total Purchase Value in between age groups.  This did not mean, however, that each person in age group 20-24 purchased the most expensive or highest value which are those on ages 35-39.

Item ID 178, Oathbreaker, Last Hope of the Breaking Storm, is the most profitable and popular among all items that were purchased.